performant-software/dm-2
# Batch image upload improvements

### What this PR does

- Per #426:
  - Creates a dialog with options for folders and progress bars on image upload
  - Combines "create doc, add image, add tile source" into single action
  - Fixes issue with thumbnail-setting timeouts
  - Warns user on close page or dialog during upload (with new dependency `react-beforeunload`)

### Additional notes

Users with non-Heroku installs will need to run
```sh
cd client && yarn install
```
to get the new dependency `react-beforeunload`.

### Status


### Steps to test

1. Visit https://dm-2-staging.herokuapp.com/
2. Log in
3. Open or create a project with Write or Admin access
4. Click "New Item > Images (batch)"
5. Do step 6 multiple times, once with each of these configurations:
  - "Place all uploads into a folder" unchecked
  - "Place all uploads into a folder" checked, "New Folder" checked (and name the folder)
  - "Place all uploads into a folder" checked, "Existing Folder" checked (and choose a folder)
6. Click "Upload Multiple" and choose many images to upload
7. At least once, try closing the page and/or closing the dialog while the uploads are in progress, and verify that a confirmation dialog appears
8. Verify that, after a few minutes, uploads either complete successfully or have a clear error message*
9. Check the uploads in your TOC / folders to verify that all looks correct

*I repeatedly ran into an error setting the thumbnail on my local instance. That may happen and I'm not entirely sure why just yet. If you run into that error, then this PR can still be accepted (in my opinion!). I'm still troubleshooting it and will treat it separately if it shows up on Heroku. If you run into another error, please mark this as rejected.
